apps/SparseMatrix/sparse_matrix.cuh

```diff
@@ -0,0 +1,191 @@
+            uint16_t s_local_id = s_ids.second;
+            col_idx[row_ptr[patch_ptr_v[patch_id] + local_id] + v + 1] =
+                patch_ptr_v[s_patch_id] + s_local_id;
+        }
+    };
+
+    query_block_dispatcher<Op::VV, blockThreads>(context, init_lambda);
+}
+
+// Follow the idea of "All calculations and storage is done on the GPU." This is
+// for initial mem allocation. This is currently VV implementation, will bge
+// extended
+template <typename T>
+void sparse_mat_init(RXMeshStatic& rx,
+                     uint32_t*&    patch_ptr_v,
+                     uint32_t*&    row_ptr,
+                     uint32_t*&    col_idx,
+                     uint32_t&     entry_size,
+                     T*&           val)
+{
+    using namespace rxmesh;
+    constexpr uint32_t blockThreads = 256;
+
+    uint32_t num_patches  = rx.get_num_patches();
+    uint32_t num_vertices = rx.get_num_vertices();
+    uint32_t num_edges    = rx.get_num_edges();
+
+    // row pointer allocation and init with prefix sum for CRS
+    CUDA_ERROR(
+        cudaMalloc((void**)&row_ptr, (num_vertices + 1) * sizeof(uint32_t)));
+    
+    CUDA_ERROR(cudaMemset(row_ptr, 0, (num_vertices + 1) * sizeof(uint32_t)));
+
+    LaunchBox<blockThreads> launch_box;
+    rx.prepare_launch_box(
+        {Op::VV}, launch_box, (void*)sparse_mat_prescan<blockThreads>);
+
+    sparse_mat_prescan<blockThreads>
+        <<<launch_box.blocks,
+           launch_box.num_threads,
+           launch_box.smem_bytes_dyn>>>(rx.get_context(), patch_ptr_v, row_ptr);
+
+    // prefix sum using CUB. 
+    void*  d_cub_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    cub::DeviceScan::ExclusiveSum(
+        d_cub_temp_storage, temp_storage_bytes, row_ptr, row_ptr, num_vertices + 1);
+    CUDA_ERROR(cudaMalloc((void**)&d_cub_temp_storage, temp_storage_bytes));
+
+    cub::DeviceScan::ExclusiveSum(
+        d_cub_temp_storage, temp_storage_bytes, row_ptr, row_ptr, num_vertices + 1);
+
+    CUDA_ERROR(cudaFree(d_cub_temp_storage));
+
+    // get entry size
+    CUDA_ERROR(cudaMemcpy(&entry_size,
+                          (row_ptr + num_vertices),
+                          sizeof(uint32_t),
+                          cudaMemcpyDeviceToHost));
+
+    // printf("%" PRIu32 " - %" PRIu32 " \n", entry_size, 2 * num_edges + num_vertices);
+
+    // column index allocation and init
+    CUDA_ERROR(cudaMalloc((void**)&col_idx, entry_size * sizeof(uint32_t)));
+    rx.prepare_launch_box(
+        {Op::VV}, launch_box, (void*)sparse_mat_col_fill<blockThreads>);
+
+    sparse_mat_col_fill<blockThreads><<<launch_box.blocks,
+                                        launch_box.num_threads,
+                                        launch_box.smem_bytes_dyn>>>(
+        rx.get_context(), patch_ptr_v, row_ptr, col_idx);
+
+    // val pointer allocation, actual value init should be in another function
+    CUDA_ERROR(cudaMalloc((void**)&val, entry_size * sizeof(uint32_t)));
+}
```

Comment:
Since all these implementation should not be exposed to the user, I would wrap them inside `namespace detail`

Reply:
Done and Can you tell me more about the namespace structure used in RXMesh? Why do wo need a detail to separate the user operation from internal? Is there more similar namespace design in RXMesh?

Reply:
It is just a convention to isolate internal implementation that user won't/shouldn't need to access https://stackoverflow.com/a/26546780/1608232

Reply:
Got it. Didn't expect it to be that simple:)